### DISCUSSION TITLE
Error when going both ways on a one-to-many relationship using async

### DISCUSSION THREAD

**Discussion by morty** [sqlalchemy/sqlalchemy#8433] [Usage Questions] [answered]
I know that async is marked as beta. I think this is a bug, but it might just be I'm doing something wrong.

It's similar to #7934 but I'm setting eager loading on the relationships to avoid having to specify at load time.

If I define a one-to-many relationship with relationships defined on both models using `back_populates` and `lazy="selectin"` when I load an instance of the many table (`Bar` in the code below), use the relationship to get the instance of the one table (`Foo` in the code below) and then try to get the collection of the many instance it fails when using async as if it was trying to do a lazy load at acccess time. Access to instances of another one-to-many relationship (`Baz` in the code below) is not effected.

```python
# models.py - run this file to generate the tables in the database and some test data
from sqlalchemy import Integer, Column, ForeignKey
from sqlalchemy.orm import declarative_base, relationship

Base = declarative_base()


class Foo(Base):
    __tablename__ = "foo"

    id = Column(Integer, primary_key=True)

    bars = relationship("Bar", back_populates="foo", uselist=True, lazy="selectin")
    bazs = relationship("Baz", back_populates="foo", uselist=True, lazy="selectin")

    def __repr__(self):
        return f"Foo(id={self.id})"


class Bar(Base):
    __tablename__ = "bar"

    id = Column(Integer, primary_key=True)
    foo_id = Column(Integer, ForeignKey("foo.id"), nullable=False)

    foo = relationship("Foo", back_populates="bars", uselist=False, lazy="selectin")

    def __repr__(self):
        return f"Bar(id={self.id})"


class Baz(Base):
    __tablename__ = "baz"

    id = Column(Integer, primary_key=True)
    foo_id = Column(Integer, ForeignKey("foo.id"), nullable=False)

    foo = relationship("Foo", back_populates="bazs", uselist=False, lazy="selectin")

    def __repr__(self):
        return f"Baz(id={self.id})"


if __name__ == "__main__":
    from sqlalchemy import create_engine
    from sqlalchemy.orm import Session

    engine = create_engine(
        "postgresql+psycopg2://scott:tiger@localhost/scott", future=True
    )
    Base.metadata.create_all(engine)

    with Session(engine) as session:
        foo = Foo(id=1)
        session.add(foo)

        bar1 = Bar(id=1, foo_id=1)
        bar2 = Bar(id=2, foo_id=1)
        session.add(bar1)
        session.add(bar2)

        baz1 = Baz(id=1, foo_id=1)
        baz2 = Baz(id=2, foo_id=1)
        session.add(baz1)
        session.add(baz2)

        session.commit()
```

```python
# sync.py
from sqlalchemy import create_engine
from sqlalchemy.orm import Session

from models import Bar

engine = create_engine("postgresql+psycopg2://scott:tiger@localhost/scott", future=True)

with Session(engine) as session:
    bar1 = session.get(Bar, 1)

    print(bar1.foo)
    print(bar1.foo.bazs)
    print(bar1.foo.bars)
```

Running this sync version outputs the following:

```
Foo(id=1)
[Baz(id=1), Baz(id=2)]
[Bar(id=1), Bar(id=2)]
```

However when I try the same thing using async I get an error `sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called` when the final print is called:

```python
# async.py
import asyncio
from models import Foo, Bar
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker, selectinload


async def main():
    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/scott", future=True
    )
    async_session = sessionmaker(engine, expire_on_commit=False, class_=AsyncSession)
    async with async_session() as session:
        options = [] #  [selectinload(Bar.foo).selectinload(Foo.bars)]
        bar1 = await session.get(Bar, 1, options=options)

        print(bar1.foo)
        print(bar1.foo.bazs)
        print(bar1.foo.bars)


if __name__ == "__main__":
    asyncio.run(main())
```

Uncommenting the `options` causes it to work as expected but I would prefer not to have to do this if possible?

**Comment by zzzeek** [comment, marked as answer]
> I know that async is marked as beta. I think this is a bug, but it might just be I'm doing something wrong.

just so you know, the workings of ORM loading are not modified at all for asyncio.   So if you are observing lazy loading happening at X or Y, that's not a "bug in asyncio" because asyncio does not implement any ORM loading behavior by itself.   It's just some loading behavior in the regular ORM that is undesirable (and may be a bug, although there are very, very few places in ORM loading that any new and unexpected bugs might occur at this point).  This is why we have a whole [section](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession) in the asyncio docs regarding ways to catch where lazyloading might occur.

> ```python
> # sync.py
> from sqlalchemy import create_engine
> from sqlalchemy.orm import Session
> 
> from models import Bar
> 
> engine = create_engine("postgresql+psycopg2://scott:tiger@localhost/scott", future=True)
> 
> with Session(engine) as session:
>     bar1 = session.get(Bar, 1)
> 
>     print(bar1.foo)
>     print(bar1.foo.bazs)
>     print(bar1.foo.bars)
> ```
> 
> Running this sync version outputs the following:
> 
> ```
> Foo(id=1)
> [Baz(id=1), Baz(id=2)]
> [Bar(id=1), Bar(id=2)]
> ```

right but it's lazy loading on that last foo.bars collection.   set echo=True on create_engine.   behavior is no different in sync vs. async.

What you are looking for here is for the selectinload loader strategy to act **recursively**, which it does not do automatically.     This means that we are loading a Bar object, then some other things load related to the Bar, then a selectinloader is enabled that is told, "load these Bar objects, from an ultimate ancestor that is also a Bar".   selectinloader says, "nope, I dont want to keep loading recursively, I dont want to blow myself out and potentially load in the entire database", which is something that could happen if there are say many-to-many relationships in the chain of loads that form an interconnected graph.   It's also not always desirable when loading an immediate relationship, that it would be going out and loading as deep as it can go, that's actually not a desirable behavior for cases where applications aren't traversing recursive graphs like that, and in SQLAlchemy most relationships are already recursive because everyone uses backrefs.

so one thing about selectinloader if you don't have many-to-many's in there which can blow things out is that it can actually be invoked arbitrarily deep without having an endless loop because it will eventually run out of objects that weren't loaded.    This is not the case for a different eagerloader, joinedload, which has to render a single SELECT statement and for which arbitrary recursiveness is not possible.  So in 2.0 we have begun to address the feature of "implicit" recursiveness, that is a selectinloader that will just automatically, without you asking, go out and keep loading until it can't anymore, that is #8126, but it right now only works for self-referential relationships.  because actually, while the idea of that is pretty simple and I had a POC working within a few minutes, actually making it work for real took a few days.

if you noticed I'm carefully using phrases like "automatically" and "without you asking" is that you can already, for any eager loader, set the "depth" that it will use when it goes out to load - this will not load arbitrarily deeply, but it will load a few levels deeply for you with no problem.   This is the [join_depth](https://docs.sqlalchemy.org/en/14/orm/relationship_api.html#sqlalchemy.orm.relationship.params.join_depth) option and in this case you can set it to 2 in order to load that deeply in all cases.  change your program as:

```py
class Foo(Base):
    __tablename__ = "foo"

    id = Column(Integer, primary_key=True)

    bars = relationship("Bar", back_populates="foo", uselist=True, lazy="selectin", join_depth=2)
    bazs = relationship("Baz", back_populates="foo", uselist=True, lazy="selectin")

    def __repr__(self):
        return f"Foo(id={self.id})"
```

I think it is likely worth a mention of this at the asyncio loading section at https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession and the general join_depth docs at https://docs.sqlalchemy.org/en/14/orm/self_referential.html#self-referential-eager-loading need some cleanup , and also should likely be somehow included in the general "loader strategy" documentation.  the whole "self referential" section is stale in various ways, I am doing a large doc change for 2.0.

**Reply by morty** [reply]
Thank you @zzzeek for the prompt and detailed explanation.
